src/hooks/useItemSelector.ts

```diff
@@ -0,0 +1,40 @@
+import { ChangeEvent, useState } from "react";
+
+const useItemSelector = <Item extends { id: number | string }>(items: Item[]) => {
+  const [selectedItems, setSelectedItems] = useState<Set<string>>(new Set());
+
+  const onToggleItem = (e: ChangeEvent<HTMLInputElement>) => {
+    const { target } = e;
+
+    const newSelectedItems = new Set(selectedItems);
+
+    selectedItems.has(target.id) ? newSelectedItems.delete(target.id) : newSelectedItems.add(target.id);
+
+    setSelectedItems(newSelectedItems);
```

Comment:
기존 state에서 update를 할 경우 콜백을 사용해주세요.
react의 setState는 동기적이지 않습니다.

```suggestion
    setSelectedItems((prevSet) => {
      const newSet = new Set(prevSet)
      prevSet.has(target.id) ? newSet.delete(target.id) : newSet.add(target.id)
      return newSet
    })
```